### PR TITLE
[LibOS] Refactor IPC subsystem's PID, SYS-V, and range logic

### PR DESCRIPTION
**Pull request by dimakuv**
## Description of the changes 

This PR contains the following changes:
- Removing macros like CONCAT, NS_SEND, NS_CALLBACK and replacing with actual function/symbol names.
- Removing distinction between PID and SYS-V namespaces and their corresponding leaders: all range requests for both namespaces now go through the same logic and are served by the same set of functions like `ipc_lease_send`/`ipc_lease_callback`, etc.
- Removing unnecessary IPC header files, by consolidating all declarations in shim_ipc.h.
- Moving logic from shim_ipc_ns.h and shim_ipc_nsimpl.h to a new shim_ipc_ranges.c -- responsible for alloc/dealloc of ranges.

## How to test this PR? 

All tests must pass. No functional change.

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)

### CODE REVIEW

**Comment by dimakuv** [comment]
Jenkins, retest this please (run 2; previous run successful)

**Comment by dimakuv** [comment]
Jenkins, retest this please (run 3; previous run successful)

**Comment by bigdata-memory** [comment]
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1597 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG63XirCOa6UbDVjG0p:-MG63XirCOa6UbDVjG0q:b-opunms) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1597)):*
> ```C
>     if (!nstatus) {
>         free(status_buf);
>         return 0;
> ```

add ```*status = NULL;``` right before return

**Comment by bigdata-memory** [comment]
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1448 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64HpP-wfa-ILNYitd:-MG64HpP-wfa-ILNYite:b-qfs1qu) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1448)):*
> ```C
> }
> 
> int get_all_pid_status(struct pid_status** status) {
> ```

check the ```status != NULL``` since it will be de-referenced later.

**Comment by bigdata-memory** [comment]
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1604 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64rdr74fFLjkUKMvB:-MG64rdr74fFLjkUKMvC:b-ijd1u2) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1604)):*
> ```C
> }
> 
> int sysv_add_key(struct sysv_key* key, IDTYPE id) {
> ```

return it immediately in case key == NULL;

**Comment by bigdata-memory** [comment]
___
*[LibOS/shim/src/ipc/shim_ipc_sysv.c, line 67 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG65Yad5LzjQ04rkVUt:-MG65Yad5LzjQ04rkVUu:b-a1il2d) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_sysv.c#L67)):*
> ```C
> 
>     ret = sysv_get_key(&msgin->key, false);
>     if (ret < 0)
> ```

remove its containing ```goto ```
```
if (ret >=  0)
  ret = ipc_sysv_tellkey_send(port, msg->src, &msgin->key, ret, msg->seq);
return ret;
```

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 16 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: all files reviewed, 10 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv)

___
*[LibOS/shim/include/shim_ipc.h, line 22 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MFzN_m5EhpXykLR4H9I:-MFzN_m647yA7mEQBYMt:b-wkclzp) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L22)):*
> ```objc
> #include <shim_types.h>
> 
> /* if callback return RESPONSE_CALLBACK, send a response even if the callback succeeded */
> ```

I think it should either be `returned` or `returns`
___
*[LibOS/shim/include/shim_ipc.h, line 39 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG0oxrIANUIXDW28Cv5:-MG0oxrIANUIXDW28Cv6:b-l37tsh) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L39)):*
> ```objc
>     IPC_DIRPRT,    /* direct parent */
>     IPC_CLT,       /* pid/sysv namespace client */
>     IPC_LDR,       /* pid/sysv namespace leader */
> ```

Please use longer names, DIRCLD, DIRPRT, CLT and LDR are hard to decipher.
Or is it too much for this PR?
___
*[LibOS/shim/include/shim_ipc.h, line 156 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG0q9Z5BxrPmK4WvijY:-MG0q9Z6-B8OR3CG9Gfx:b-8w8pjg) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L156)):*
> ```
> allocate_ipc
> release_ipc
> ```
I'd suffix both of them with `_id`, otherwise it sounds like allocating some kind of IPC object, not just reserving an ID.
___
*[LibOS/shim/include/shim_ipc.h, line 163 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG0qZXf5GF5qrkurUOm:-MG0qZXf5GF5qrkurUOn:bhr66zf) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L163)):*
> ```objc
> /* sysv namespace */
> #define KEY_HASH(k)      ((k)->key)
> #define KEY_COMP(k1, k2) ((k1)->key != (k2)->key || (k1)->type != (k2)->type)
> ```

`KEY_COMP` -> `KEY_IS_EQUAL` + flip its returned value

I know this was already here before this PR, but it triggers me too much :) Right now this macro pretends to be a typical C comparator, but it returns bool with an unintuitive meaning (so, it's neither -1/0/1 nor true==equal/false==not equal)
___
*[LibOS/shim/include/shim_ipc.h, line 247 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64B_M28OsA-S9mWz4:-MG64B_M28OsA-S9mWz5:b-klpmfm) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L247)):*
> ```objc
> int ipc_renew_callback(struct shim_ipc_msg* msg, struct shim_ipc_port* port);
> 
> /* SUBLEASE: lease a range of names */
> ```

why do we call them "names"? aren't they just numeric IDs?
___
*[LibOS/shim/include/shim_ipc.h, line 269 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64fQC6Ew-KMotH-5X:-MG64fQC6Ew-KMotH-5Y:bplo6a3) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L269)):*
> ```objc
> int ipc_queryall_callback(struct shim_ipc_msg* msg, struct shim_ipc_port* port);
> 
> /* ANSWER: answer the channel of certain names */
> ```

I know they were here before, but I have troubles parsing these comments grammatically. What's the intended meaning?
___
*[LibOS/shim/include/shim_ipc.h, line 271 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64vOR85z49-LIqSdt:-MG64vOR85z49-LIqSdu:b4tmhnw) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/include/shim_ipc.h#L271)):*
> ```objc
> /* ANSWER: answer the channel of certain names */
> struct shim_ipc_answer {
>     int nanswers;
> ```

while we're here, could you change it to `size_t` and rename to `answers_cnt`?
___
*[LibOS/shim/src/ipc/shim_ipc.c, line 69 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG667i21FGfVniysv3S:-MG667i39CxlclxfEv_p:b-4ul5qj) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc.c#L69)):*
> ```C
> }
> 
> int prepare_ns_leaders(void) {
> ```

do we still need this function?
___
*[LibOS/shim/src/ipc/shim_ipc_helper.c, line 62 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG66ZoSBprMwIaRGR4C:-MG66ZoSBprMwIaRGR4D:b-bh3rkj) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_helper.c#L62)):*
> ```C
>     /* IPC_QUERY        */ &ipc_query_callback,
>     /* IPC_QUERYALL     */ &ipc_queryall_callback,
>     /* IPC_ANSWER       */ &ipc_answer_callback,
> ```

While changing this, could you switch to `[KEY] = VALUE` initializer syntax?
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1448 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64HpP-wfa-ILNYitd:-MG8M6Ry1svezTgp2_NZ:b-foa0ic) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1448)):*
<details><summary><i>Previously, bigdata-memory (Gordon King) wrote…</i></summary><blockquote>

check the ```status != NULL``` since it will be de-referenced later.
</blockquote></details>

no need to, IMO
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1453 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG8OaT53az0O1_46d4G:-MG8OaT53az0O1_46d4H:b-aicra8) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1453)):*
> ```C
> 
>     int bufsize                   = RANGE_SIZE;
>     struct pid_status* status_buf = malloc(bufsize);
> ```

While we're here, could you check malloc result here?
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1604 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64rdr74fFLjkUKMvB:-MG8QUmn7sHxoI3HdNVn:b6fyruz) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1604)):*
<details><summary><i>Previously, bigdata-memory (Gordon King) wrote…</i></summary><blockquote>

return it immediately in case key == NULL;
</blockquote></details>

I don't think `key == NULL` is legal here, why would it be?

**Comment by mkow** [comment]
Jenkins, retest this please (run 4; previous run successful)

**Comment by dimakuv** [comment]
Jenkins, retest Jenkins please (`kill01` LTP test failed)

**Review by mkow — COMMENTED** [review]
> Reviewed 13 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1780#-MGAF_id8kbs1h3Dl8TC:-MGAwMwhAp6GyfgRh5rO:beux1ss) (no related file):*
<details><summary><i>Previously, yamahata wrote…</i></summary><blockquote>

Just to check my understanding, pid and sysv share same id space with this change.
i.e. we can't have same id for both pid and sysv.
How does get_pid_all_ranges() list only pids? it's used for "/proc/pid" emulation.

</blockquote></details>

Hmm, good question...

---
___
*[LibOS/shim/include/shim_ipc.h, line 161 at r2](https://reviewable.io/reviews/oscarlab/graphene/1780#-MGA3q9_AUwF0WN58AOb:-MGAwkEd9QUwN5pI7_TF:b32v57) ([raw file](https://github.com/oscarlab/graphene/blob/def20ac34bb5fd5379f9b64016841f6401fc09fb/LibOS/shim/include/shim_ipc.h#L161)):*
<details><summary><i>Previously, yamahata wrote…</i></summary><blockquote>

Now this is only used for sysv_key only. So we can safely convert this into inline function instead of macro.
</blockquote></details>

+1
___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1604 at r1](https://reviewable.io/reviews/oscarlab/graphene/1780#-MG64rdr74fFLjkUKMvB:-MGAyEDV5RQP8WqwcMwO:b-ump8g) ([raw file](https://github.com/oscarlab/graphene/blob/b4425db69848b1b7153b0bd7099c5c5f2ce4d999/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1604)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Done. Added an assert. Indeed `key == NULL` is not supposed to happen here.
</blockquote></details>

What I mean is that it doesn't make sense to check all arguments in all functions for validity - we should do this only on public interfaces or for things which aren't intuitive or easy to use incorrectly. In this particular case passing NULL will most likely crash Graphene, so we're not risking slipping a silent bug.
The default state is that if a function does accept a pointer it can't be NULL unless explicitly documented that this argument it's optional.

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @yamahata)

___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1629 at r3](https://reviewable.io/reviews/oscarlab/graphene/1780#-MGBH2efACeo0GM1x9Oh:-MGBH2efACeo0GM1x9Oi:bwpu2ql) ([raw file](https://github.com/oscarlab/graphene/blob/587a26012bed07269a9d5631ac4cf407b413833f/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1629)):*
> ```C
>     k->key.key  = key->key;
>     k->key.type = key->type;
>     k->id        = id;
> ```

broken alignment

**Review by yamahata — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1780#-MGAF_id8kbs1h3Dl8TC:-MGCYbuXCuUrSsNqZzWl:b-yyumec) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Good question. Your understanding is correct -- PID and SYS-V resources share the same ID space now. We cannot have the same ID for both PID and SYS-V resources.

So when we emulate `/proc/pid`, it calls `get_all_pid_status()` to collect all known resource IDs like this:
1. `get_all_pid_status(&status)` to put all collected info about known PIDs in the `status` array
2. Call `ipc_queryall_send()` to collect all info about all known IDs (both PID and SYS-V) in the `offered_ranges` array
3.  For each range, call `ipc_pid_getstatus_send(<ipc-port-to-owner-process>, &range_status)` to ask the owner of the range to return array of PID statuses in the `range_status` array
    - Each owner receives this `IPC_PID_GETSTATUS` message and reacts via `ipc_pid_getstatus_callback()`
    - `ipc_pid_getstatus_callback()` walks the owner's thread list via `shim_ipc_pid.c: check_thread()` and **crucially** performs the comparison `status->pids[i] == thread->tid`
    - at this point, SYS-V IDs are filtered out, and the only IDs left are the PID ones, allocated for process's threads
    - The owner calls `ipc_pid_retstatus_send(&status)` which sends the array of only PIDs back
4. All PIDs from all owners are collected in the final `&status` array by repeating step 3
5. Done

TLDR: Even though the logic consolidates both PIDs and SYS-V ranges, there is a filtering step to only collect PIDs.
</blockquote></details>

Thanks for detailed explanation.

---

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: 18 of 19 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @mkow and @yamahata)

___
*[LibOS/shim/src/ipc/shim_ipc_ranges.c, line 1629 at r3](https://reviewable.io/reviews/oscarlab/graphene/1780#-MGBH2efACeo0GM1x9Oh:-MGFMzcCEvWltuVGyP2T:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/587a26012bed07269a9d5631ac4cf407b413833f/LibOS/shim/src/ipc/shim_ipc_ranges.c#L1629)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

broken alignment
</blockquote></details>

Done.

**Review by mkow — DISMISSED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: all files reviewed, all discussions resolved, "fixup! " found in commit messages' one-liners

**Review by yamahata — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: all files reviewed, all discussions resolved, not enough approvals from different teams (1 more required, approved so far: Intel)

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1780)* status: :shipit: complete! all files reviewed, all discussions resolved
